recipes/libffi/3.3/conanfile.py

```diff
@@ -0,0 +1,193 @@
+                "LIBTOOL": tools.unix_path(os.path.join(self.source_folder, self._source_subfolder, "ltmain.sh")),
+                "CC": msvcc,
+                "CXX": msvcc,
+                "LD": "link",
+                "CPP": "cl -nologo -EP",
+                "CXXCPP": "cl -nologo -EP",
+            })
+        with tools.environment_append(extra_env_vars):
+            yield
+
+    def build(self):
+        self._patch_files()
+        config_args = [
+            "--enable-debug" if self.settings.build_type == "Debug" else "--disable-debug",
+            "--enable-shared" if self.options.shared else "--disable-shared",
+            "--disable-static" if self.options.shared else "--enable-static",
+            "--disable-multi-os-directory",
+        ]
+        autotools = self._get_auto_tools()
+        build = None
+        host = None
+        if self.settings.compiler == "Visual Studio":
+            build = "{}-{}-{}".format(
+                "x86_64" if "64" in platform.machine() else "i686",
+                "pc" if self.settings.arch == "x86" else "w64",
+                "cygwin")
+            host = "{}-{}-{}".format(
+                "x86_64" if self.settings.arch == "x86_64" else "i686",
+                "pc" if self.settings.arch == "x86" else "w64",
+                "cygwin")
+        else:
+            if autotools.host and "x86-" in autotools.host:
+                autotools.host = autotools.host.replace("x86", "i686")
+        with self._create_auto_tools_environment(autotools):
+            autotools.configure(configure_dir=os.path.join(self.source_folder, self._source_subfolder),
+                                build=build,
+                                host=host,
+                                args=config_args)
+            autotools.make()
+            if tools.get_env("CONAN_RUN_TESTS", False):
+                autotools.make(target="check")
+
+    def package(self):
+        self.copy("LICENSE", src=os.path.join(self.source_folder, self._source_subfolder), dst="licenses")
+        if self.settings.os == "Windows":
+            self.copy("*.h", src="{}/include".format(self.build_folder), dst="include")
+        if self.settings.compiler == "Visual Studio":
+            self.copy("*.lib", src="{}/.libs".format(self.build_folder), dst="lib")
+            self.copy("*.dll", src="{}/.libs".format(self.build_folder), dst="bin")
+        else:
+            autotools = self._get_auto_tools()
+            with self._create_auto_tools_environment(autotools):
+                autotools.install()
+            os.unlink(os.path.join(self.package_folder, "lib", "libffi.la"))
+            tools.rmdir(os.path.join(self.package_folder, "lib", "pkgconfig"))
+            tools.rmdir(os.path.join(self.package_folder, "share"))
+
+    def package_info(self):
+        self.cpp_info.includedirs = ["include"]
+        if self.options.shared:
+            self.cpp_info.defines += ["FFI_BUILDING"]
```

Comment:
probably, it should be `FF_SHARED`/`FFI_STATIC` or something like that (like in 3.2.1)?
usually, `XXX_BUILDING` macro is defined while library itself is being built, not while it's being consumed.

Reply:
See `source_subfolder/include/ffi.h.in`:
```
/* Need minimal decorations for DLLs to work on Windows.  GCC has
   autoimport and autoexport.  Always mark externally visible symbols
   as dllimport for MSVC clients, even if it means an extra indirection
   when using the static version of the library.
   Besides, as a workaround, they can define FFI_BUILDING if they
   *know* they are going to link with the static library.  */
#if defined _MSC_VER
# if defined FFI_BUILDING_DLL /* Building libffi.DLL with msvcc.sh */
#  define FFI_API __declspec(dllexport)
# elif !defined FFI_BUILDING  /* Importing libffi.DLL */
#  define FFI_API __declspec(dllimport)
# else                        /* Building/linking static library */
#  define FFI_API
# endif
#else
# define FFI_API
#endif
```

So I was incorrect, `FFI_BUILDING` must be defined when doing a static build.